{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "**NOTE**\n",
    "\n",
    "This document was generated by converting the `rsem2hbadeals.R` script to .ipynb using the python library [`jupytext`](https://github.com/mwouts/jupytext), by typing the following command:\n",
    "\n",
    "\n",
    "```bash\n",
    "jupytext --to ipynb rsem2hbadeals.R\n",
    "```\n",
    "You should see something like the following in your console after typing the above command:\n",
    "\n",
    "```\n",
    "[jupytext] Reading rsem2hbadeals.R\n",
    "[jupytext] Writing rsem2hbadeals.ipynb\n",
    "```\n",
    "\n",
    "The above tells you your `.ipynb` version of your initial script has been created. You can now open it and continue editting your notebooks in a Notebook Session. \n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is the .ipynb version of `rsem2hbadeals.R`\n",
    "\n",
    "The script was created to be used as a Nextflow process, to run HBA-DEALS on the output of RSEM, added to the peer reviewed and community developed pipeline [`nf-core/rnaseq`](https://github.com/nf-core/rnaseq). The script can be found here: <br>\n",
    "https://github.com/cgpu/HBA-DEALS/blob/cgpu-rsem2hbadeals/docker/rsem2hbadeals.R\n",
    "\n",
    "\n",
    "The respective Nextflow process can be found here: <br>\n",
    "https://github.com/TheJacksonLaboratory/coronavirus-nf/blob/master/main.nf#L1645-L1673\n",
    "\n",
    "```groovy\n",
    "\n",
    "    process hbadeals {\n",
    "            tag \"${contrast_id}\"\n",
    "            label \"hbadeals\"\n",
    "            publishDir \"${params.outdir}/hbadeals\", mode: \"${params.publish_dir_mode}\"\n",
    "\n",
    "            input:\n",
    "                set val(contrast_id), file(metadata) from ch_hbadeals_metadata\n",
    "                file(\"*\") from rsem_results_isoforms_hbadeals.collect()\n",
    "\n",
    "            output:\n",
    "                file(\"${contrast_id}.csv\") into hbadeals_results_isoforms\n",
    "\n",
    "            when:\n",
    "            !params.skip_rsem && !params.skip_hbadeals\n",
    "\n",
    "\n",
    "            script:\n",
    "            \"\"\"\n",
    "            rsem2hbadeals.R \\\n",
    "            --rsem_folder='.' \\\n",
    "            --metadata=$metadata \\\n",
    "            --rsem_file_suffix=$params.rsem_file_suffix \\\n",
    "            --output=$contrast_id \\\n",
    "            --isoform_level=$params.isoform_level \\\n",
    "            --mcmc_iter=$params.mcmc_iter \\\n",
    "            --mcmc_warmup=$params.mcmc_warmup \\\n",
    "            --n_cores=${task.cpus}\n",
    "            \"\"\"\n",
    "    }\n",
    "```\n",
    "\n",
    "\n",
    "The default parameters of `rsem2hbadeals.R` for testing are defined in the configuration Nextflow specific file named [`nextflow.config`](https://github.com/cgpu/rnaseq/blob/2ea4af7f964e5adbee7bf95f46af12de1e6d7094/nextflow.config#L85-L90) and are exposing all the arguments of the function [`hbadeals`](https://github.com/TheJacksonLaboratory/HBA-DEALS/blob/master/R/hbadeals.R) of the R package [HBA-DEALS](https://github.com/TheJacksonLaboratory/HBA-DEALS).\n",
    "\n",
    "```groovy\n",
    "  accessionList     = false               // required, A file must be provided, will be used for countsData\n",
    "  rsem_file_suffix  = '.isoforms.results' // optional, default set here\n",
    "  isoform_level     = 'TRUE'              // optional, default set here\n",
    "  mcmc_iter         = 20                  // optional, default set here\n",
    "  mcmc_warmup       = 10                  // optional, default set here\n",
    "  hbadeals_metadata = false               // required, A file must be provided, will be used for labels\n",
    "```\n",
    "\n",
    "See below the documentation of the function `hbadeals::hbadeals()` for detailed information on expected input arguments:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for hbadeals {hbadeals}\"><tr><td>hbadeals {hbadeals}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>\n",
       "Hierarchical Bayesian Analysis of Differential Expression and ALternative SPlicing (HBA-DEALS)\n",
       "</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>The function determines differential expression and transcript proportions between two experimental conditions.  It takes as input a table that contains the gene names in its first column, the transcript names in its second column,\n",
       "and the corresponding transcript counts under every sample in the rest of its columns.  In addition, it takes as input an ordered numeric vector of 1's and 2's, where 1 indicates that the corresponding sample belongs to the first experimental condition and 2 to the second. The function returns a table that provides 1-probability values for differential expression of each gene, and either differential splicing for the gene or differential proportion for each transcript.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "hbadeals(countsData,labels,n.cores,isoform.level,mcmc.iter,mcmc.warmup)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>countsData</code></td>\n",
       "<td>\n",
       "\n",
       "<p>A table of gene names, transcript names and transcript counts in each sample.  At least two transcripts must correspond to\n",
       "each gene.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>labels</code></td>\n",
       "<td>\n",
       "\n",
       "<p>An ordered vector of 1's and 2's.  Its length is ncol(countsdata)-2.  Each entry indicates whether the corresponding sample/column\n",
       "of countsData belongs to the first experimental condition or the second.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>n.cores</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The number of cores to use in the calculation.  It is recommended to dedicate as many cores as possible.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>isoform.level</code></td>\n",
       "<td>\n",
       "\n",
       "<p>Should 1-probability of differential proportion for each transcript be returned.  FALSE by deafult, which returns 1-probability\n",
       "of differential splicing for each gene.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mcmc.iter</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The number of iterations of the MCMC algorithm after warmup.  The recommended value is the default.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mcmc.warmup</code></td>\n",
       "<td>\n",
       "\n",
       "<p>The number of warmup iterations of the MCMC algorithm.  The recommended value is the default.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>The table returned by this function contains 4 columns.  The first column is the gene name, the second is the transcript name,\n",
       "the third is the fold change, and the fourth is 1-probability of differential expression or proportion.  Entries that\n",
       "refer to expression have 'Expression' in their second column.  If isoform.level is FALSE, entries that refer to differential splicing of the gene will have 'Splicing' in their second column entry.  The fold change for expression is given as log fold change, and for splicing\n",
       "as fold change.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Guy Karlebach, Peter N Robinson\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "simulate(rseed=1,fc=3,equal = TRUE)\n",
       "n.samples=4 #per condition\n",
       "countsData=read.table('counts.txt',sep='\\t',header=FALSE)\n",
       "labels=c(rep(1,n.samples),rep(2,n.samples))\n",
       "res=hbadeals(countsData,labels)\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>hbadeals</em> version 0.99.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{hbadeals}{Hierarchical Bayesian Analysis of Differential Expression and ALternative SPlicing (HBA-DEALS)}{hbadeals}\n",
       "\\keyword{\\textasciitilde{}expression}{hbadeals}\n",
       "\\keyword{\\textasciitilde{}splicing}{hbadeals}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "The function determines differential expression and transcript proportions between two experimental conditions.  It takes as input a table that contains the gene names in its first column, the transcript names in its second column,\n",
       "and the corresponding transcript counts under every sample in the rest of its columns.  In addition, it takes as input an ordered numeric vector of 1's and 2's, where 1 indicates that the corresponding sample belongs to the first experimental condition and 2 to the second. The function returns a table that provides 1-probability values for differential expression of each gene, and either differential splicing for the gene or differential proportion for each transcript.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "hbadeals(countsData,labels,n.cores,isoform.level,mcmc.iter,mcmc.warmup)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{countsData}] \n",
       "A table of gene names, transcript names and transcript counts in each sample.  At least two transcripts must correspond to\n",
       "each gene.\n",
       "\n",
       "\\item[\\code{labels}] \n",
       "An ordered vector of 1's and 2's.  Its length is ncol(countsdata)-2.  Each entry indicates whether the corresponding sample/column\n",
       "of countsData belongs to the first experimental condition or the second.\n",
       "\n",
       "\\item[\\code{n.cores}] \n",
       "The number of cores to use in the calculation.  It is recommended to dedicate as many cores as possible.\n",
       "\n",
       "\\item[\\code{isoform.level}] \n",
       "Should 1-probability of differential proportion for each transcript be returned.  FALSE by deafult, which returns 1-probability\n",
       "of differential splicing for each gene.\n",
       "\n",
       "\\item[\\code{mcmc.iter}] \n",
       "The number of iterations of the MCMC algorithm after warmup.  The recommended value is the default.\n",
       "\n",
       "\\item[\\code{mcmc.warmup}] \n",
       "The number of warmup iterations of the MCMC algorithm.  The recommended value is the default.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "The table returned by this function contains 4 columns.  The first column is the gene name, the second is the transcript name,\n",
       "the third is the fold change, and the fourth is 1-probability of differential expression or proportion.  Entries that\n",
       "refer to expression have 'Expression' in their second column.  If isoform.level is FALSE, entries that refer to differential splicing of the gene will have 'Splicing' in their second column entry.  The fold change for expression is given as log fold change, and for splicing\n",
       "as fold change.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Guy Karlebach, Peter N Robinson\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "simulate(rseed=1,fc=3,equal = TRUE)\n",
       "n.samples=4 #per condition\n",
       "countsData=read.table('counts.txt',sep='\\t',header=FALSE)\n",
       "labels=c(rep(1,n.samples),rep(2,n.samples))\n",
       "res=hbadeals(countsData,labels)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "hbadeals               package:hbadeals                R Documentation\n",
       "\n",
       "_\bH_\bi_\be_\br_\ba_\br_\bc_\bh_\bi_\bc_\ba_\bl _\bB_\ba_\by_\be_\bs_\bi_\ba_\bn _\bA_\bn_\ba_\bl_\by_\bs_\bi_\bs _\bo_\bf _\bD_\bi_\bf_\bf_\be_\br_\be_\bn_\bt_\bi_\ba_\bl _\bE_\bx_\bp_\br_\be_\bs_\bs_\bi_\bo_\bn _\ba_\bn_\bd\n",
       "_\bA_\bL_\bt_\be_\br_\bn_\ba_\bt_\bi_\bv_\be _\bS_\bP_\bl_\bi_\bc_\bi_\bn_\bg (_\bH_\bB_\bA-_\bD_\bE_\bA_\bL_\bS)\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     The function determines differential expression and transcript\n",
       "     proportions between two experimental conditions.  It takes as\n",
       "     input a table that contains the gene names in its first column,\n",
       "     the transcript names in its second column, and the corresponding\n",
       "     transcript counts under every sample in the rest of its columns.\n",
       "     In addition, it takes as input an ordered numeric vector of 1's\n",
       "     and 2's, where 1 indicates that the corresponding sample belongs\n",
       "     to the first experimental condition and 2 to the second. The\n",
       "     function returns a table that provides 1-probability values for\n",
       "     differential expression of each gene, and either differential\n",
       "     splicing for the gene or differential proportion for each\n",
       "     transcript.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     hbadeals(countsData,labels,n.cores,isoform.level,mcmc.iter,mcmc.warmup)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "countsData: A table of gene names, transcript names and transcript\n",
       "          counts in each sample.  At least two transcripts must\n",
       "          correspond to each gene.\n",
       "\n",
       "  labels: An ordered vector of 1's and 2's.  Its length is\n",
       "          ncol(countsdata)-2.  Each entry indicates whether the\n",
       "          corresponding sample/column of countsData belongs to the\n",
       "          first experimental condition or the second.\n",
       "\n",
       " n.cores: The number of cores to use in the calculation.  It is\n",
       "          recommended to dedicate as many cores as possible.\n",
       "\n",
       "isoform.level: Should 1-probability of differential proportion for each\n",
       "          transcript be returned.  FALSE by deafult, which returns\n",
       "          1-probability of differential splicing for each gene.\n",
       "\n",
       "mcmc.iter: The number of iterations of the MCMC algorithm after warmup.\n",
       "          The recommended value is the default.\n",
       "\n",
       "mcmc.warmup: The number of warmup iterations of the MCMC algorithm.\n",
       "          The recommended value is the default.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     The table returned by this function contains 4 columns.  The first\n",
       "     column is the gene name, the second is the transcript name, the\n",
       "     third is the fold change, and the fourth is 1-probability of\n",
       "     differential expression or proportion.  Entries that refer to\n",
       "     expression have 'Expression' in their second column.  If\n",
       "     isoform.level is FALSE, entries that refer to differential\n",
       "     splicing of the gene will have 'Splicing' in their second column\n",
       "     entry.  The fold change for expression is given as log fold\n",
       "     change, and for splicing as fold change.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Guy Karlebach, Peter N Robinson\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     simulate(rseed=1,fc=3,equal = TRUE)\n",
       "     n.samples=4 #per condition\n",
       "     countsData=read.table('counts.txt',sep='\\t',header=FALSE)\n",
       "     labels=c(rep(1,n.samples),rep(2,n.samples))\n",
       "     res=hbadeals(countsData,labels)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?hbadeals::hbadeals()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The `rsem2hbadeals.R` helper function:\n",
    "\n",
    "Type the following command in your terminal to inspect the required arguments:\n",
    "\n",
    "```bash\n",
    "./rsem2hbadeals --help\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "      The helper R Script rsem2hbadeals.R\n",
    "      Mandatory arguments:\n",
    "          --rsem_folder=path        - The path to the folder with the RSEM output files. The expected filenames from RSEM are:\n",
    "                                      '<sample_id>.isoforms.results', '<sample_id>.genes.results'\n",
    "                                      The 'transcript_id', 'gene_id' and 'expected_count' columns MUST be present in the files generated by RSEM.\n",
    "                                      '.' is accepted for denoting that the files are in the current working directory.\n",
    "\n",
    "          --metadata=path           - A comma seperated file with metadata about the samples in the RSEM files.\n",
    "                                      Rownames must be the samples. Column names must be metadata about the samples.\n",
    "                                      Two columns MUST be present in the file. One with the SRR id and one that denotes the control/case status.\n",
    "                                      The default column names are for the two mandatory columsn are:\n",
    "                                      1. 'sample_id': accepted values are valid SRR ids eg. SRR10503939\n",
    "                                      2. 'status': accepted values are one of {'control', 'case'}\n",
    "                                      Optionally, alternative column names can be defined by the user see --sample_colname, --status_colname below.\n",
    "                                      NOTE: The SRR ids must include the samples that correspond to the RSEM files provided via --rsem_folder.\n",
    "\n",
    "          --rsem_file_suffix=value  - A string in single quotes, one of the following:\n",
    "                                      'isoform.results' or 'genes.results'.\n",
    "\n",
    "          --output=path             - A filename to save the hbadeals::hbadeals() output.\n",
    "\n",
    "         --help                     - you are reading it\n",
    "\n",
    "      Optionnal arguments:\n",
    "          --sample_colname=chr      - Column name in the metadata file that contains the SRR id, default: 'sample_id'\n",
    "          --status_colname=chr      - Column name in the metadata file with the control,case status information, default: 'status'\n",
    "          --n_cores=value           - Number of cores for hbadeals::hbadeals(), default: 2\n",
    "          --isoform_level=boolean   - TRUE or FALSE. Check ?hbadeals::hbadeals, default: TRUE\n",
    "          --mcmc_warmup=value       - Number of warmup iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 20\n",
    "          --mcmc_iter=value         - Number of iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 100\n",
    "\n",
    "     Usage:\n",
    "\n",
    "          The typical command for running the script is as follows:\n",
    "\n",
    "          ./rsem2hbadeals.R --rsem_folder='.' --metadata='./metadata.csv' --rsem_file_suffix='.isoforms.results' --output='hbadeals_results.csv'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reproducing an `rstan` error message \n",
    "\n",
    "This error occcured when running `hbadeals::hbadeals()` with 2 controls and 2 cases.<br>\n",
    "The error can be inspected in the CloudOS analysis job page here:<br>\n",
    "https://cloudos.lifebit.ai/public/jobs/5ea2ba865d1cd5010348bbcf\n",
    "\n",
    "\n",
    "To view the whole error message for the failed `hbadeals` Nextflow process:\n",
    "\n",
    "1. visit the link https://cloudos.lifebit.ai/public/jobs/5ea2ba865d1cd5010348bbcf\n",
    "2. click on **`FAILED`**\n",
    "3. click on **`Show log`**\n",
    "\n",
    "See the steps described above in the following gif:\n",
    "\n",
    " <img src=\"http://g.recordit.co/0GNdXC5GN4.gif\"  width=\"650\"/></a>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Passing same arguments as for the rsem2hbadeals.R script in the Nextflow pipeline:\n",
    "\n",
    "rsem_folder        <- '../data'\n",
    "metadata           <- '../data/metadata.csv'\n",
    "sample_colname     <- 'sample_id'\n",
    "status_colname     <- 'status'\n",
    "rsem_file_suffix   <- '.isoforms.results'\n",
    "output             <- '2vs2_samples'\n",
    "n_cores            <- '2'\n",
    "isoform_level      <- TRUE\n",
    "mcmc_warmup        <- 20\n",
    "mcmc_iter          <- 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ARGUMENTS SUMMARY\n",
      "rsem_folder      : ../data\n",
      "metadata         : ../data/metadata.csv\n",
      "sample_colname   : sample_id\n",
      "status_colname   : status\n",
      "rsem_file_suffix : .isoforms.results\n",
      "output           : 2vs2_samples\n",
      "n_cores          : 2\n",
      "isoform_level    : TRUE\n",
      "mcmc_warmup      : 20\n",
      "mcmc_iter        : 10\n"
     ]
    }
   ],
   "source": [
    "cat(\"\\n\")\n",
    "cat(\"ARGUMENTS SUMMARY\")\n",
    "cat(\"\\n\")\n",
    "cat(\"rsem_folder      : \", rsem_folder,    \"\\n\",sep=\"\")\n",
    "cat(\"metadata         : \", metadata,        \"\\n\",sep=\"\")\n",
    "cat(\"sample_colname   : \", sample_colname,  \"\\n\",sep=\"\")\n",
    "cat(\"status_colname   : \", status_colname,  \"\\n\",sep=\"\")\n",
    "cat(\"rsem_file_suffix : \", rsem_file_suffix,\"\\n\",sep=\"\")\n",
    "cat(\"output           : \", output          ,\"\\n\",sep=\"\")\n",
    "cat(\"n_cores          : \", n_cores,         \"\\n\",sep=\"\")\n",
    "cat(\"isoform_level    : \", isoform_level,   \"\\n\",sep=\"\")\n",
    "cat(\"mcmc_warmup      : \", mcmc_warmup,     \"\\n\",sep=\"\")\n",
    "cat(\"mcmc_iter        : \", mcmc_iter,       \"\\n\",sep=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################# LIBRARIES SECTION #############################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressWarnings(suppressMessages(library(limma)))\n",
    "suppressWarnings(suppressMessages(library(coda)))\n",
    "suppressWarnings(suppressMessages(library(rstan)))\n",
    "suppressWarnings(suppressMessages(library(codetools)))\n",
    "suppressWarnings(suppressMessages(library(hbadeals)))\n",
    "\n",
    "# Not needed in main rsem2hbadeals.R script - Needed only for the error reproducible example \n",
    "suppressWarnings(suppressMessages(library(piggyback)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "############################### SCRIPT SECTION ###############################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieving the input data for the reproducible example\n",
    "I have included here the input data as a compressed archive `.tar.gz`:<br>\n",
    "https://github.com/cgpu/HBA-DEALS/releases/tag/rstan-error\n",
    "\n",
    " <img src=\"http://g.recordit.co/IqHxuDITXP.gif\"  width=\"650\"/></a>\n",
    "\n",
    "Copy the command from the releases to fetch the data and store them temporarily in the `HBA-DEALS/data` folder:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The required archive 'reprex_rstan_error.tar.gz' is already available in the folder ../data.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!(file.exists('../data/SRR10503929.isoforms.results'))) {\n",
    "    piggyback::pb_download(\n",
    "    repo = \"cgpu/HBA-DEALS\", \n",
    "    file = \"reprex_rstan_error.tar.gz\", \n",
    "    tag  = \"rstan-error\",\n",
    "    dest = '../data')\n",
    "} else {\n",
    "    message(\"The required archive 'reprex_rstan_error.tar.gz' is already available in the folder ../data.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The archive 'reprex_rstan_error.tar.gz' has already been decompressed.\n",
      "Required files ('*isoforms.results', 'metadata.csv') are available in the folder ../data.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Decompress, untar archive:\n",
    "\n",
    "if (!(file.exists('../data/SRR10503929.isoforms.results'))) {\n",
    "system( 'tar -xzvf ../data/reprex_rstan_error.tar.gz -C ../data', intern = TRUE)\n",
    "} else {\n",
    "    message(\"The archive 'reprex_rstan_error.tar.gz' has already been decompressed.\\nRequired files ('*isoforms.results', 'metadata.csv') are available in the folder ../data.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'countsData_2vs2_samples.csv'</li><li>'datalist'</li><li>'metadata.csv'</li><li>'reprex_rstan_error.tar.gz'</li><li>'SRR10503928.isoforms.results'</li><li>'SRR10503929.isoforms.results'</li><li>'SRR10503936.isoforms.results'</li><li>'SRR10503937.isoforms.results'</li><li>'voom_sim.RData'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'countsData\\_2vs2\\_samples.csv'\n",
       "\\item 'datalist'\n",
       "\\item 'metadata.csv'\n",
       "\\item 'reprex\\_rstan\\_error.tar.gz'\n",
       "\\item 'SRR10503928.isoforms.results'\n",
       "\\item 'SRR10503929.isoforms.results'\n",
       "\\item 'SRR10503936.isoforms.results'\n",
       "\\item 'SRR10503937.isoforms.results'\n",
       "\\item 'voom\\_sim.RData'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'countsData_2vs2_samples.csv'\n",
       "2. 'datalist'\n",
       "3. 'metadata.csv'\n",
       "4. 'reprex_rstan_error.tar.gz'\n",
       "5. 'SRR10503928.isoforms.results'\n",
       "6. 'SRR10503929.isoforms.results'\n",
       "7. 'SRR10503936.isoforms.results'\n",
       "8. 'SRR10503937.isoforms.results'\n",
       "9. 'voom_sim.RData'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"countsData_2vs2_samples.csv\"  \"datalist\"                    \n",
       "[3] \"metadata.csv\"                 \"reprex_rstan_error.tar.gz\"   \n",
       "[5] \"SRR10503928.isoforms.results\" \"SRR10503929.isoforms.results\"\n",
       "[7] \"SRR10503936.isoforms.results\" \"SRR10503937.isoforms.results\"\n",
       "[9] \"voom_sim.RData\"              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Verify the files are in the folder HBA-DEALS/data:\n",
    "list.files('../data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start of commands for running HBA-DEALS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read metadata and select required columns\n",
    "accepted_status_labels_chr <- c(\"control\", \"case\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "metaData  <- read.table(metadata, header = TRUE, sep = \",\")\n",
    "metaData  <- metaData[, c(sample_colname, status_colname)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 4 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample_id</th><th scope=col>status</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SRR10503928</td><td>control</td></tr>\n",
       "\t<tr><td>SRR10503929</td><td>control</td></tr>\n",
       "\t<tr><td>SRR10503936</td><td>case   </td></tr>\n",
       "\t<tr><td>SRR10503937</td><td>case   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 2\n",
       "\\begin{tabular}{ll}\n",
       " sample\\_id & status\\\\\n",
       " <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t SRR10503928 & control\\\\\n",
       "\t SRR10503929 & control\\\\\n",
       "\t SRR10503936 & case   \\\\\n",
       "\t SRR10503937 & case   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 2\n",
       "\n",
       "| sample_id &lt;fct&gt; | status &lt;fct&gt; |\n",
       "|---|---|\n",
       "| SRR10503928 | control |\n",
       "| SRR10503929 | control |\n",
       "| SRR10503936 | case    |\n",
       "| SRR10503937 | case    |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_id   status \n",
       "1 SRR10503928 control\n",
       "2 SRR10503929 control\n",
       "3 SRR10503936 case   \n",
       "4 SRR10503937 case   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Validate metadata columns for required specs\n",
    "if (!(length(unique(metaData[[status_colname]]))) == 2){\n",
    "    stop(\"The status column in the metadata table has more than 2 levels (eg. case, control, something_else).\\nType ./rsem2hbadeals.R --help to check the metadata specifications.\")\n",
    "}\n",
    "if(! (all(unique(metaData[[status_colname]]) %in% accepted_status_labels_chr)) ) {\n",
    "    stop(\"The status column in the metadata table has values other {'control','case'}, which are the only ones accepted.\\nType ./rsem2hbadeals.R --help to check the metadata specifications.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to control,case to 1,2 as requested in hbadeals::hbadeals()\n",
    "metaData[[status_colname]] <- ifelse( metaData[[status_colname]] == \"control\", 1, 2)\n",
    "colnames(metaData)  <- c(\"sample_id\", \"status\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 4 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample_id</th><th scope=col>status</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SRR10503928</td><td>1</td></tr>\n",
       "\t<tr><td>SRR10503929</td><td>1</td></tr>\n",
       "\t<tr><td>SRR10503936</td><td>2</td></tr>\n",
       "\t<tr><td>SRR10503937</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 2\n",
       "\\begin{tabular}{ll}\n",
       " sample\\_id & status\\\\\n",
       " <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t SRR10503928 & 1\\\\\n",
       "\t SRR10503929 & 1\\\\\n",
       "\t SRR10503936 & 2\\\\\n",
       "\t SRR10503937 & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 2\n",
       "\n",
       "| sample_id &lt;fct&gt; | status &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| SRR10503928 | 1 |\n",
       "| SRR10503929 | 1 |\n",
       "| SRR10503936 | 2 |\n",
       "| SRR10503937 | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_id   status\n",
       "1 SRR10503928 1     \n",
       "2 SRR10503929 1     \n",
       "3 SRR10503936 2     \n",
       "4 SRR10503937 2     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metaData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'../data/SRR10503928.isoforms.results'</li><li>'../data/SRR10503929.isoforms.results'</li><li>'../data/SRR10503936.isoforms.results'</li><li>'../data/SRR10503937.isoforms.results'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '../data/SRR10503928.isoforms.results'\n",
       "\\item '../data/SRR10503929.isoforms.results'\n",
       "\\item '../data/SRR10503936.isoforms.results'\n",
       "\\item '../data/SRR10503937.isoforms.results'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '../data/SRR10503928.isoforms.results'\n",
       "2. '../data/SRR10503929.isoforms.results'\n",
       "3. '../data/SRR10503936.isoforms.results'\n",
       "4. '../data/SRR10503937.isoforms.results'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"../data/SRR10503928.isoforms.results\"\n",
       "[2] \"../data/SRR10503929.isoforms.results\"\n",
       "[3] \"../data/SRR10503936.isoforms.results\"\n",
       "[4] \"../data/SRR10503937.isoforms.results\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get paths of RSEM files\n",
    "isoform_individual_files <- list.files(rsem_folder,\n",
    "                                       full.names = TRUE,\n",
    "                                       pattern    = rsem_file_suffix)\n",
    "isoform_individual_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep 'gene_id' and 'transcript_id' columns only (scaffold to build collective countsData csv from all samples)\n",
    "countsData  <- read.table(isoform_individual_files[1], header=TRUE)\n",
    "countsData  <- countsData[, c(\"transcript_id\", \"gene_id\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000373020.9_TSPAN6-201</td><td>ENSG00000000003.15_TSPAN6</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000494424.1_TSPAN6-202</td><td>ENSG00000000003.15_TSPAN6</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000496771.5_TSPAN6-203</td><td>ENSG00000000003.15_TSPAN6</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000612152.4_TSPAN6-204</td><td>ENSG00000000003.15_TSPAN6</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000614008.4_TSPAN6-205</td><td>ENSG00000000003.15_TSPAN6</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000373031.5_TNMD-201  </td><td>ENSG00000000005.6_TNMD   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & transcript\\_id & gene\\_id\\\\\n",
       "  & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000373020.9\\_TSPAN6-201 & ENSG00000000003.15\\_TSPAN6\\\\\n",
       "\t2 & ENST00000494424.1\\_TSPAN6-202 & ENSG00000000003.15\\_TSPAN6\\\\\n",
       "\t3 & ENST00000496771.5\\_TSPAN6-203 & ENSG00000000003.15\\_TSPAN6\\\\\n",
       "\t4 & ENST00000612152.4\\_TSPAN6-204 & ENSG00000000003.15\\_TSPAN6\\\\\n",
       "\t5 & ENST00000614008.4\\_TSPAN6-205 & ENSG00000000003.15\\_TSPAN6\\\\\n",
       "\t6 & ENST00000373031.5\\_TNMD-201   & ENSG00000000005.6\\_TNMD   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | transcript_id &lt;fct&gt; | gene_id &lt;fct&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENST00000373020.9_TSPAN6-201 | ENSG00000000003.15_TSPAN6 |\n",
       "| 2 | ENST00000494424.1_TSPAN6-202 | ENSG00000000003.15_TSPAN6 |\n",
       "| 3 | ENST00000496771.5_TSPAN6-203 | ENSG00000000003.15_TSPAN6 |\n",
       "| 4 | ENST00000612152.4_TSPAN6-204 | ENSG00000000003.15_TSPAN6 |\n",
       "| 5 | ENST00000614008.4_TSPAN6-205 | ENSG00000000003.15_TSPAN6 |\n",
       "| 6 | ENST00000373031.5_TNMD-201   | ENSG00000000005.6_TNMD    |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id                gene_id                  \n",
       "1 ENST00000373020.9_TSPAN6-201 ENSG00000000003.15_TSPAN6\n",
       "2 ENST00000494424.1_TSPAN6-202 ENSG00000000003.15_TSPAN6\n",
       "3 ENST00000496771.5_TSPAN6-203 ENSG00000000003.15_TSPAN6\n",
       "4 ENST00000612152.4_TSPAN6-204 ENSG00000000003.15_TSPAN6\n",
       "5 ENST00000614008.4_TSPAN6-205 ENSG00000000003.15_TSPAN6\n",
       "6 ENST00000373031.5_TNMD-201   ENSG00000000005.6_TNMD   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(countsData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iteratively append each individual sample RSEM 'expected_count' column to create cohort table\n",
    "for (input.file in isoform_individual_files  ) {\n",
    "    next.file  <- read.table(input.file, header=TRUE)\n",
    "    sample_id  <- gsub(rsem_file_suffix, \"\", basename(input.file))\n",
    "    \n",
    "    # Heuristic to name a column after the sample id\n",
    "    next.file[[sample_id ]] <- next.file$expected_count\n",
    "    toKeep     <- c(\"transcript_id\", sample_id)\n",
    "    countsData <- merge(countsData, next.file[, toKeep], by = \"transcript_id\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>SRR10503928</th><th scope=col>SRR10503929</th><th scope=col>SRR10503936</th><th scope=col>SRR10503937</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000000233.10_ARF5-201  </td><td>ENSG00000004059.11_ARF5   </td><td> 436.26</td><td> 443.20</td><td>281.70</td><td>257.44</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000000412.8_M6PR-201   </td><td>ENSG00000003056.8_M6PR    </td><td> 924.00</td><td> 885.68</td><td>536.73</td><td>585.91</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000000442.11_ESRRA-201 </td><td>ENSG00000173153.15_ESRRA  </td><td> 236.07</td><td> 195.35</td><td> 84.03</td><td> 88.92</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000001008.6_FKBP4-201  </td><td>ENSG00000004478.8_FKBP4   </td><td>1965.40</td><td>1985.91</td><td>896.15</td><td>845.11</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENST00000001146.6_CYP26B1-201</td><td>ENSG00000003137.8_CYP26B1 </td><td>   0.00</td><td>   7.06</td><td>  8.21</td><td>  0.00</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000002125.9_NDUFAF7-201</td><td>ENSG00000003509.16_NDUFAF7</td><td>  54.37</td><td>  66.25</td><td> 33.80</td><td> 46.75</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & transcript\\_id & gene\\_id & SRR10503928 & SRR10503929 & SRR10503936 & SRR10503937\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000000233.10\\_ARF5-201   & ENSG00000004059.11\\_ARF5    &  436.26 &  443.20 & 281.70 & 257.44\\\\\n",
       "\t2 & ENST00000000412.8\\_M6PR-201    & ENSG00000003056.8\\_M6PR     &  924.00 &  885.68 & 536.73 & 585.91\\\\\n",
       "\t3 & ENST00000000442.11\\_ESRRA-201  & ENSG00000173153.15\\_ESRRA   &  236.07 &  195.35 &  84.03 &  88.92\\\\\n",
       "\t4 & ENST00000001008.6\\_FKBP4-201   & ENSG00000004478.8\\_FKBP4    & 1965.40 & 1985.91 & 896.15 & 845.11\\\\\n",
       "\t5 & ENST00000001146.6\\_CYP26B1-201 & ENSG00000003137.8\\_CYP26B1  &    0.00 &    7.06 &   8.21 &   0.00\\\\\n",
       "\t6 & ENST00000002125.9\\_NDUFAF7-201 & ENSG00000003509.16\\_NDUFAF7 &   54.37 &   66.25 &  33.80 &  46.75\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | transcript_id &lt;fct&gt; | gene_id &lt;fct&gt; | SRR10503928 &lt;dbl&gt; | SRR10503929 &lt;dbl&gt; | SRR10503936 &lt;dbl&gt; | SRR10503937 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000000233.10_ARF5-201   | ENSG00000004059.11_ARF5    |  436.26 |  443.20 | 281.70 | 257.44 |\n",
       "| 2 | ENST00000000412.8_M6PR-201    | ENSG00000003056.8_M6PR     |  924.00 |  885.68 | 536.73 | 585.91 |\n",
       "| 3 | ENST00000000442.11_ESRRA-201  | ENSG00000173153.15_ESRRA   |  236.07 |  195.35 |  84.03 |  88.92 |\n",
       "| 4 | ENST00000001008.6_FKBP4-201   | ENSG00000004478.8_FKBP4    | 1965.40 | 1985.91 | 896.15 | 845.11 |\n",
       "| 5 | ENST00000001146.6_CYP26B1-201 | ENSG00000003137.8_CYP26B1  |    0.00 |    7.06 |   8.21 |   0.00 |\n",
       "| 6 | ENST00000002125.9_NDUFAF7-201 | ENSG00000003509.16_NDUFAF7 |   54.37 |   66.25 |  33.80 |  46.75 |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id                 gene_id                    SRR10503928\n",
       "1 ENST00000000233.10_ARF5-201   ENSG00000004059.11_ARF5     436.26    \n",
       "2 ENST00000000412.8_M6PR-201    ENSG00000003056.8_M6PR      924.00    \n",
       "3 ENST00000000442.11_ESRRA-201  ENSG00000173153.15_ESRRA    236.07    \n",
       "4 ENST00000001008.6_FKBP4-201   ENSG00000004478.8_FKBP4    1965.40    \n",
       "5 ENST00000001146.6_CYP26B1-201 ENSG00000003137.8_CYP26B1     0.00    \n",
       "6 ENST00000002125.9_NDUFAF7-201 ENSG00000003509.16_NDUFAF7   54.37    \n",
       "  SRR10503929 SRR10503936 SRR10503937\n",
       "1  443.20     281.70      257.44     \n",
       "2  885.68     536.73      585.91     \n",
       "3  195.35      84.03       88.92     \n",
       "4 1985.91     896.15      845.11     \n",
       "5    7.06       8.21        0.00     \n",
       "6   66.25      33.80       46.75     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Inspect updated countsData\n",
    "head(countsData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sample size = 4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculating sample size\n",
    "sample_size <-  dim(countsData)[2] - 2\n",
    "message(paste0('Sample size = ', sample_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SRR10503928'</li><li>'SRR10503929'</li><li>'SRR10503936'</li><li>'SRR10503937'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'SRR10503928'\n",
       "\\item 'SRR10503929'\n",
       "\\item 'SRR10503936'\n",
       "\\item 'SRR10503937'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'SRR10503928'\n",
       "2. 'SRR10503929'\n",
       "3. 'SRR10503936'\n",
       "4. 'SRR10503937'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"SRR10503928\" \"SRR10503929\" \"SRR10503936\" \"SRR10503937\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Temporarily dettach metadata columns \"transcript_id\",\"gene_id\" from numerical values\n",
    "all      <- colnames(countsData)\n",
    "toRemove <- c(\"transcript_id\",\"gene_id\")\n",
    "toKeep   <- all [!all  %in% toRemove]\n",
    "toKeep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Applying exclusion criterion 1:  ( discard transcripts with very low count across samples) ..\n",
      "\n",
      "Done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Apply exclusion criterion 1: Filter out transcripts with cumsum across samples lower than the N, cohort size\n",
    "message(\"\\nApplying exclusion criterion 1:  ( discard transcripts with very low count across samples) ..\")\n",
    "countsData <- countsData[rowSums(countsData[,toKeep] > 0 ) >= sample_size, ]\n",
    "message(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Applying exclusion criterion 2: ( keep > 1 isoform transcripts) ..\n",
      "\n",
      "Done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Apply exclusion criterion 2: Filter out transcripts with only one isoform\n",
    "message(\"\\nApplying exclusion criterion 2: ( keep > 1 isoform transcripts) ..\")\n",
    "num.iso=unlist(lapply(countsData$gene_id, function(x){sum(countsData$gene_id %in% x)}))\n",
    "countsData <- countsData[num.iso > 1, ]\n",
    "message(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Ordering 'countsData' columns according to 'metaData$sample_id' (labels) order ..\n",
      "\n",
      "Done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Reorder countsData columns by metaData order !DANGEROUS TO RELY ON INDEXES!\n",
    "message(\"\\nOrdering 'countsData' columns according to 'metaData$sample_id' (labels) order ..\")\n",
    "toKeepInOrder <- c(\"transcript_id\", \"gene_id\", as.vector(metaData$sample_id) )\n",
    "countsData    <- countsData[, toKeepInOrder]\n",
    "message(\"Done!\")\n",
    "write.table(countsData,\n",
    "            file = paste0(\"../data/countsData_\", output, \".csv\"),\n",
    "            sep       =',',\n",
    "            quote     = F,\n",
    "            col.names = T,\n",
    "            row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'numeric'"
      ],
      "text/latex": [
       "'numeric'"
      ],
      "text/markdown": [
       "'numeric'"
      ],
      "text/plain": [
       "[1] \"numeric\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li><li>2</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "3. 2\n",
       "4. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 1 2 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>SRR10503928</th><th scope=col>SRR10503929</th><th scope=col>SRR10503936</th><th scope=col>SRR10503937</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000000233.10_ARF5-201</td><td>ENSG00000004059.11_ARF5</td><td>436.26</td><td>443.2</td><td>281.7</td><td>257.44</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & transcript\\_id & gene\\_id & SRR10503928 & SRR10503929 & SRR10503936 & SRR10503937\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000000233.10\\_ARF5-201 & ENSG00000004059.11\\_ARF5 & 436.26 & 443.2 & 281.7 & 257.44\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 6\n",
       "\n",
       "| <!--/--> | transcript_id &lt;fct&gt; | gene_id &lt;fct&gt; | SRR10503928 &lt;dbl&gt; | SRR10503929 &lt;dbl&gt; | SRR10503936 &lt;dbl&gt; | SRR10503937 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000000233.10_ARF5-201 | ENSG00000004059.11_ARF5 | 436.26 | 443.2 | 281.7 | 257.44 |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id               gene_id                 SRR10503928 SRR10503929\n",
       "1 ENST00000000233.10_ARF5-201 ENSG00000004059.11_ARF5 436.26      443.2      \n",
       "  SRR10503936 SRR10503937\n",
       "1 281.7       257.44     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(as.numeric(metaData$status))\n",
    "as.numeric(metaData$status)\n",
    "head(countsData, 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the simulated data from the HBA-DEALS package to compare R objects `class()`\n",
    "\n",
    "> **Note**: The code for simulating data can be found in the [HBA-DEALS vignette](https://github.com/TheJacksonLaboratory/HBA-DEALS/blob/master/vignettes/vignette.Rmd)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "hbadeals::simulate(rseed=1,fc=3,equal = T)\n",
    "n.samples=4\n",
    "sim_countsData=read.table('counts.txt',sep='\\t',header=F)\n",
    "labels=c(rep(1,n.samples),rep(2,n.samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li><li>1</li><li>1</li><li>2</li><li>2</li><li>2</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "3. 1\n",
       "4. 1\n",
       "5. 2\n",
       "6. 2\n",
       "7. 2\n",
       "8. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 1 1 1 2 2 2 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'numeric'"
      ],
      "text/latex": [
       "'numeric'"
      ],
      "text/markdown": [
       "'numeric'"
      ],
      "text/plain": [
       "[1] \"numeric\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th><th scope=col>V7</th><th scope=col>V8</th><th scope=col>V9</th><th scope=col>V10</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Gene-258</td><td>iso-1</td><td>1</td><td>2</td><td>2</td><td>3</td><td>2</td><td>2</td><td>2</td><td>3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & V1 & V2 & V3 & V4 & V5 & V6 & V7 & V8 & V9 & V10\\\\\n",
       "  & <fct> & <fct> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & Gene-258 & iso-1 & 1 & 2 & 2 & 3 & 2 & 2 & 2 & 3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 10\n",
       "\n",
       "| <!--/--> | V1 &lt;fct&gt; | V2 &lt;fct&gt; | V3 &lt;int&gt; | V4 &lt;int&gt; | V5 &lt;int&gt; | V6 &lt;int&gt; | V7 &lt;int&gt; | V8 &lt;int&gt; | V9 &lt;int&gt; | V10 &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Gene-258 | iso-1 | 1 | 2 | 2 | 3 | 2 | 2 | 2 | 3 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1       V2    V3 V4 V5 V6 V7 V8 V9 V10\n",
       "1 Gene-258 iso-1 1  2  2  3  2  2  2  3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "labels\n",
    "class(labels)\n",
    "head(sim_countsData, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in regularize.values(x, y, ties, missing(ties)):\n",
      "“collapsing to unique 'x' values”\n",
      "recompiling to avoid crashing R session\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Time difference of 5.562764 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run HBA-DEALS on the simulated data\n",
    "# message('\\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')\n",
    "start_time <- Sys.time()\n",
    "res_sim <-hbadeals::hbadeals(countsData     = sim_countsData,\n",
    "                              labels        = labels,\n",
    "                              n.cores       = parallel::detectCores(),\n",
    "                              isoform.level = TRUE,\n",
    "                              mcmc.iter     = mcmc_iter,\n",
    "                              mcmc.warmup   = mcmc_warmup)\n",
    "end_time <- Sys.time()\n",
    "runtime <- end_time - start_time\n",
    "runtime\n",
    "# message('Successfully completed hbadeals::hbadeals() task!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 6 × 4 of type chr</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene</th><th scope=col>Isoform</th><th scope=col>ExplogFC/FC</th><th scope=col>P</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Gene-258</td><td>Expression</td><td>0.923111630567003 </td><td>0.1</td></tr>\n",
       "\t<tr><td>Gene-258</td><td>iso-1     </td><td>0.710325756415805 </td><td>0.1</td></tr>\n",
       "\t<tr><td>Gene-258</td><td>iso-2     </td><td>1.28068776688241  </td><td>0.1</td></tr>\n",
       "\t<tr><td>Gene-282</td><td>Expression</td><td>-0.432707505719108</td><td>0.4</td></tr>\n",
       "\t<tr><td>Gene-282</td><td>iso-1     </td><td>0.850702341153059 </td><td>0.3</td></tr>\n",
       "\t<tr><td>Gene-282</td><td>iso-2     </td><td>1.25912910872316  </td><td>0.3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 4 of type chr\n",
       "\\begin{tabular}{llll}\n",
       " Gene & Isoform & ExplogFC/FC & P\\\\\n",
       "\\hline\n",
       "\t Gene-258 & Expression & 0.923111630567003  & 0.1\\\\\n",
       "\t Gene-258 & iso-1      & 0.710325756415805  & 0.1\\\\\n",
       "\t Gene-258 & iso-2      & 1.28068776688241   & 0.1\\\\\n",
       "\t Gene-282 & Expression & -0.432707505719108 & 0.4\\\\\n",
       "\t Gene-282 & iso-1      & 0.850702341153059  & 0.3\\\\\n",
       "\t Gene-282 & iso-2      & 1.25912910872316   & 0.3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 4 of type chr\n",
       "\n",
       "| Gene | Isoform | ExplogFC/FC | P |\n",
       "|---|---|---|---|\n",
       "| Gene-258 | Expression | 0.923111630567003  | 0.1 |\n",
       "| Gene-258 | iso-1      | 0.710325756415805  | 0.1 |\n",
       "| Gene-258 | iso-2      | 1.28068776688241   | 0.1 |\n",
       "| Gene-282 | Expression | -0.432707505719108 | 0.4 |\n",
       "| Gene-282 | iso-1      | 0.850702341153059  | 0.3 |\n",
       "| Gene-282 | iso-2      | 1.25912910872316   | 0.3 |\n",
       "\n"
      ],
      "text/plain": [
       "     Gene     Isoform    ExplogFC/FC        P  \n",
       "[1,] Gene-258 Expression 0.923111630567003  0.1\n",
       "[2,] Gene-258 iso-1      0.710325756415805  0.1\n",
       "[3,] Gene-258 iso-2      1.28068776688241   0.1\n",
       "[4,] Gene-282 Expression -0.432707505719108 0.4\n",
       "[5,] Gene-282 iso-1      0.850702341153059  0.3\n",
       "[6,] Gene-282 iso-2      1.25912910872316   0.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(res_sim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in regularize.values(x, y, ties, missing(ties)):\n",
      "“collapsing to unique 'x' values”\n",
      "Warning message in regularize.values(x, y, ties, missing(ties)):\n",
      "“collapsing to unique 'x' values”\n",
      "Warning message in mclapply(1:length(unique(countsData[, 1])), function(gene.number) {:\n",
      "“all scheduled cores encountered errors in user code”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Time difference of 26.32628 secs"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run HBA-DEALS on the real data\n",
    "# message('\\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')\n",
    "start_time <- Sys.time()\n",
    "res_real <-hbadeals::hbadeals(countsData    = countsData,\n",
    "                              labels        = as.numeric(metaData$status),\n",
    "                              n.cores       = n_cores,\n",
    "                              isoform.level = isoform_level,\n",
    "                              mcmc.iter     = mcmc_iter,\n",
    "                              mcmc.warmup   = mcmc_warmup)\n",
    "    # message('Successfully completed hbadeals::hbadeals() task!')\n",
    "end_time <- Sys.time()\n",
    "runtime <- end_time - start_time\n",
    "runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                           V1        \n",
       " Error in 1:ncol(stanFit) : NA/NaN argument\\n:27693  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(res_real)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): 'list' object cannot be coerced to type 'double'\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): 'list' object cannot be coerced to type 'double'\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "head(countsData[,toKeep]) <- as.numeric(head(countsData[,toKeep]))\n",
    "head(countsData[,toKeep])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Attempt to overcome issue: Try with 3 vs 3 control, case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>SRR10503928</th><th scope=col>SRR10503929</th><th scope=col>SRR10503936</th><th scope=col>SRR10503937</th><th scope=col>SRRfake_control</th><th scope=col>SRRfake_case</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000000233.10_ARF5-201  </td><td>ENSG00000004059.11_ARF5   </td><td> 436.26</td><td> 443.20</td><td>281.70</td><td>257.44</td><td> 436.26</td><td>257.44</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENST00000000412.8_M6PR-201   </td><td>ENSG00000003056.8_M6PR    </td><td> 924.00</td><td> 885.68</td><td>536.73</td><td>585.91</td><td> 924.00</td><td>585.91</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENST00000000442.11_ESRRA-201 </td><td>ENSG00000173153.15_ESRRA  </td><td> 236.07</td><td> 195.35</td><td> 84.03</td><td> 88.92</td><td> 236.07</td><td> 88.92</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENST00000001008.6_FKBP4-201  </td><td>ENSG00000004478.8_FKBP4   </td><td>1965.40</td><td>1985.91</td><td>896.15</td><td>845.11</td><td>1965.40</td><td>845.11</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENST00000002125.9_NDUFAF7-201</td><td>ENSG00000003509.16_NDUFAF7</td><td>  54.37</td><td>  66.25</td><td> 33.80</td><td> 46.75</td><td>  54.37</td><td> 46.75</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>ENST00000002165.11_FUCA2-201 </td><td>ENSG00000001036.14_FUCA2  </td><td> 810.07</td><td> 857.13</td><td>472.22</td><td>441.78</td><td> 810.07</td><td>441.78</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & transcript\\_id & gene\\_id & SRR10503928 & SRR10503929 & SRR10503936 & SRR10503937 & SRRfake\\_control & SRRfake\\_case\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000000233.10\\_ARF5-201   & ENSG00000004059.11\\_ARF5    &  436.26 &  443.20 & 281.70 & 257.44 &  436.26 & 257.44\\\\\n",
       "\t2 & ENST00000000412.8\\_M6PR-201    & ENSG00000003056.8\\_M6PR     &  924.00 &  885.68 & 536.73 & 585.91 &  924.00 & 585.91\\\\\n",
       "\t3 & ENST00000000442.11\\_ESRRA-201  & ENSG00000173153.15\\_ESRRA   &  236.07 &  195.35 &  84.03 &  88.92 &  236.07 &  88.92\\\\\n",
       "\t4 & ENST00000001008.6\\_FKBP4-201   & ENSG00000004478.8\\_FKBP4    & 1965.40 & 1985.91 & 896.15 & 845.11 & 1965.40 & 845.11\\\\\n",
       "\t6 & ENST00000002125.9\\_NDUFAF7-201 & ENSG00000003509.16\\_NDUFAF7 &   54.37 &   66.25 &  33.80 &  46.75 &   54.37 &  46.75\\\\\n",
       "\t7 & ENST00000002165.11\\_FUCA2-201  & ENSG00000001036.14\\_FUCA2   &  810.07 &  857.13 & 472.22 & 441.78 &  810.07 & 441.78\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | transcript_id &lt;fct&gt; | gene_id &lt;fct&gt; | SRR10503928 &lt;dbl&gt; | SRR10503929 &lt;dbl&gt; | SRR10503936 &lt;dbl&gt; | SRR10503937 &lt;dbl&gt; | SRRfake_control &lt;dbl&gt; | SRRfake_case &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000000233.10_ARF5-201   | ENSG00000004059.11_ARF5    |  436.26 |  443.20 | 281.70 | 257.44 |  436.26 | 257.44 |\n",
       "| 2 | ENST00000000412.8_M6PR-201    | ENSG00000003056.8_M6PR     |  924.00 |  885.68 | 536.73 | 585.91 |  924.00 | 585.91 |\n",
       "| 3 | ENST00000000442.11_ESRRA-201  | ENSG00000173153.15_ESRRA   |  236.07 |  195.35 |  84.03 |  88.92 |  236.07 |  88.92 |\n",
       "| 4 | ENST00000001008.6_FKBP4-201   | ENSG00000004478.8_FKBP4    | 1965.40 | 1985.91 | 896.15 | 845.11 | 1965.40 | 845.11 |\n",
       "| 6 | ENST00000002125.9_NDUFAF7-201 | ENSG00000003509.16_NDUFAF7 |   54.37 |   66.25 |  33.80 |  46.75 |   54.37 |  46.75 |\n",
       "| 7 | ENST00000002165.11_FUCA2-201  | ENSG00000001036.14_FUCA2   |  810.07 |  857.13 | 472.22 | 441.78 |  810.07 | 441.78 |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id                 gene_id                    SRR10503928\n",
       "1 ENST00000000233.10_ARF5-201   ENSG00000004059.11_ARF5     436.26    \n",
       "2 ENST00000000412.8_M6PR-201    ENSG00000003056.8_M6PR      924.00    \n",
       "3 ENST00000000442.11_ESRRA-201  ENSG00000173153.15_ESRRA    236.07    \n",
       "4 ENST00000001008.6_FKBP4-201   ENSG00000004478.8_FKBP4    1965.40    \n",
       "6 ENST00000002125.9_NDUFAF7-201 ENSG00000003509.16_NDUFAF7   54.37    \n",
       "7 ENST00000002165.11_FUCA2-201  ENSG00000001036.14_FUCA2    810.07    \n",
       "  SRR10503929 SRR10503936 SRR10503937 SRRfake_control SRRfake_case\n",
       "1  443.20     281.70      257.44       436.26         257.44      \n",
       "2  885.68     536.73      585.91       924.00         585.91      \n",
       "3  195.35      84.03       88.92       236.07          88.92      \n",
       "4 1985.91     896.15      845.11      1965.40         845.11      \n",
       "6   66.25      33.80       46.75        54.37          46.75      \n",
       "7  857.13     472.22      441.78       810.07         441.78      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "countsData$SRRfake_control <- countsData$SRR10503928\n",
    "countsData$SRRfake_case    <- countsData$SRR10503937"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample_id</th><th scope=col>status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>SRR10503928    </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>SRR10503929    </td><td>1</td></tr>\n",
       "\t<tr><th scope=row></th><td>SRRfake_control</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & sample\\_id & status\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & SRR10503928     & 1\\\\\n",
       "\t2 & SRR10503929     & 1\\\\\n",
       "\t & SRRfake\\_control & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| <!--/--> | sample_id &lt;chr&gt; | status &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | SRR10503928     | 1 |\n",
       "| 2 | SRR10503929     | 1 |\n",
       "| <!----> | SRRfake_control | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_id       status\n",
       "1 SRR10503928     1     \n",
       "2 SRR10503929     1     \n",
       "  SRRfake_control 1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "controls <- metaData[ metaData$status=='1', ]\n",
    "extra_control <- data.frame( sample_id=\"SRRfake_control\", status=1)\n",
    "controls <- as.data.frame(rbind(as.matrix(controls), as.matrix(extra_control)), stringsAsFactors = FALSE)\n",
    "controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample_id</th><th scope=col>status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>3</th><td>SRR10503936 </td><td>2</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>SRR10503937 </td><td>2</td></tr>\n",
       "\t<tr><th scope=row></th><td>SRRfake_case</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & sample\\_id & status\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t3 & SRR10503936  & 2\\\\\n",
       "\t4 & SRR10503937  & 2\\\\\n",
       "\t & SRRfake\\_case & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 2\n",
       "\n",
       "| <!--/--> | sample_id &lt;chr&gt; | status &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 3 | SRR10503936  | 2 |\n",
       "| 4 | SRR10503937  | 2 |\n",
       "| <!----> | SRRfake_case | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_id    status\n",
       "3 SRR10503936  2     \n",
       "4 SRR10503937  2     \n",
       "  SRRfake_case 2     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cases <- metaData[ metaData$status=='2', ]\n",
    "extra_case <- data.frame( sample_id=\"SRRfake_case\", status=2)\n",
    "cases <- as.data.frame(rbind(as.matrix(cases), as.matrix(extra_case)), stringsAsFactors = FALSE)\n",
    "cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample_id</th><th scope=col>status</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>SRR10503928    </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>SRR10503929    </td><td>1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>SRRfake_control</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>SRR10503936    </td><td>2</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>SRR10503937    </td><td>2</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>SRRfake_case   </td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & sample\\_id & status\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & SRR10503928     & 1\\\\\n",
       "\t2 & SRR10503929     & 1\\\\\n",
       "\t3 & SRRfake\\_control & 1\\\\\n",
       "\t4 & SRR10503936     & 2\\\\\n",
       "\t5 & SRR10503937     & 2\\\\\n",
       "\t6 & SRRfake\\_case    & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | sample_id &lt;chr&gt; | status &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | SRR10503928     | 1 |\n",
       "| 2 | SRR10503929     | 1 |\n",
       "| 3 | SRRfake_control | 1 |\n",
       "| 4 | SRR10503936     | 2 |\n",
       "| 5 | SRR10503937     | 2 |\n",
       "| 6 | SRRfake_case    | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_id       status\n",
       "1 SRR10503928     1     \n",
       "2 SRR10503929     1     \n",
       "3 SRRfake_control 1     \n",
       "4 SRR10503936     2     \n",
       "5 SRR10503937     2     \n",
       "6 SRRfake_case    2     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metaData_3vs3 <- as.data.frame(rbind(as.matrix(controls), as.matrix(cases)), stringsAsFactors = FALSE)\n",
    "rownames(metaData_3vs3) <- seq(dim(metaData_3vs3)[1])\n",
    "metaData_3vs3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Ordering 'countsData' columns according to 'metaData$sample_id' (labels) order ..\n",
      "\n",
      "Done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Reorder countsData columns by metaData order !DANGEROUS TO RELY ON INDEXES!\n",
    "message(\"\\nOrdering 'countsData' columns according to 'metaData$sample_id' (labels) order ..\")\n",
    "toKeepInOrder <- c(\"transcript_id\", \"gene_id\", as.vector(metaData_3vs3$sample_id) )\n",
    "countsData    <- countsData[, toKeepInOrder]\n",
    "message(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li><li>1</li><li>2</li><li>2</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "3. 1\n",
       "4. 2\n",
       "5. 2\n",
       "6. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 1 1 2 2 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 1 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>transcript_id</th><th scope=col>gene_id</th><th scope=col>SRR10503928</th><th scope=col>SRR10503929</th><th scope=col>SRRfake_control</th><th scope=col>SRR10503936</th><th scope=col>SRR10503937</th><th scope=col>SRRfake_case</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENST00000000233.10_ARF5-201</td><td>ENSG00000004059.11_ARF5</td><td>436.26</td><td>443.2</td><td>436.26</td><td>281.7</td><td>257.44</td><td>257.44</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & transcript\\_id & gene\\_id & SRR10503928 & SRR10503929 & SRRfake\\_control & SRR10503936 & SRR10503937 & SRRfake\\_case\\\\\n",
       "  & <fct> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENST00000000233.10\\_ARF5-201 & ENSG00000004059.11\\_ARF5 & 436.26 & 443.2 & 436.26 & 281.7 & 257.44 & 257.44\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 8\n",
       "\n",
       "| <!--/--> | transcript_id &lt;fct&gt; | gene_id &lt;fct&gt; | SRR10503928 &lt;dbl&gt; | SRR10503929 &lt;dbl&gt; | SRRfake_control &lt;dbl&gt; | SRR10503936 &lt;dbl&gt; | SRR10503937 &lt;dbl&gt; | SRRfake_case &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENST00000000233.10_ARF5-201 | ENSG00000004059.11_ARF5 | 436.26 | 443.2 | 436.26 | 281.7 | 257.44 | 257.44 |\n",
       "\n"
      ],
      "text/plain": [
       "  transcript_id               gene_id                 SRR10503928 SRR10503929\n",
       "1 ENST00000000233.10_ARF5-201 ENSG00000004059.11_ARF5 436.26      443.2      \n",
       "  SRRfake_control SRR10503936 SRR10503937 SRRfake_case\n",
       "1 436.26          281.7       257.44      257.44      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "as.numeric(metaData_3vs3$status)\n",
    "head(countsData, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in regularize.values(x, y, ties, missing(ties)):\n",
      "“collapsing to unique 'x' values”\n",
      "Warning message in regularize.values(x, y, ties, missing(ties)):\n",
      "“collapsing to unique 'x' values”\n",
      "Warning message in mclapply(1:length(unique(countsData[, 1])), function(gene.number) {:\n",
      "“all scheduled cores encountered errors in user code”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Time difference of 32.43383 secs"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run HBA-DEALS on the real data but 3 vs 3 samples control, case respectively\n",
    "# message('\\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')\n",
    "start_time <- Sys.time()\n",
    "res_real <-hbadeals::hbadeals(countsData    = countsData,\n",
    "                              labels        = as.numeric(metaData_3vs3$status),\n",
    "                              n.cores       = n_cores,\n",
    "                              isoform.level = isoform_level,\n",
    "                              mcmc.iter     = mcmc_iter,\n",
    "                              mcmc.warmup   = mcmc_warmup)\n",
    "    # message('Successfully completed hbadeals::hbadeals() task!')\n",
    "end_time <- Sys.time()\n",
    "runtime <- end_time - start_time\n",
    "runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Attempt to overcome issue: Try with 3 vs 3 control, case and random sample for the duplicated columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run HBA-DEALS on the real data\n",
    "# message('\\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')\n",
    "start_time <- Sys.time()\n",
    "res_real <-hbadeals::hbadeals(countsData    = countsData,\n",
    "                              labels        = as.numeric(metaData_3vs3$status),\n",
    "                              n.cores       = n_cores,\n",
    "                              isoform.level = isoform_level,\n",
    "                              mcmc.iter     = mcmc_iter,\n",
    "                              mcmc.warmup   = mcmc_warmup)\n",
    "    # message('Successfully completed hbadeals::hbadeals() task!')\n",
    "end_time <- Sys.time()\n",
    "runtime <- end_time - start_time\n",
    "runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write results in file\n",
    "# message('\\nWriting  hbadeals::hbadeals() results to a file..')\n",
    "write.table(res,\n",
    "            file = paste0(output),\n",
    "            sep       =',',\n",
    "            quote     = F,\n",
    "            col.names = T,\n",
    "            row.names = F)\n",
    "# message('Done!')\n",
    "# message('\\nGoodbye!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "executable": "/usr/bin/env Rscript",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
