#!/usr/bin/env Rscript

############################## ARGUMENTS SECTION #############################
## Collect arguments
args <- commandArgs(TRUE)

## Default setting when no all arguments passed or help needed
if("--help" %in% args | "help" %in% args | (length(args) == 0) | (length(args) == 1) ) {
  cat("
      The helper R Script rsem2hbadeals.R
      Mandatory arguments:
          --rsem_folder=path        - The path to the folder with the RSEM output files. The expected filenames from RSEM are:
                                      '<sample_id>.isoforms.results', '<sample_id>.genes.results'
                                      The 'transcript_id', 'gene_id' and 'expected_count' columns MUST be present in the files generated by RSEM.
                                      '.' is accepted for denoting that the files are in the current working directory.

          --metadata=path           - A comma seperated file with metadata about the samples in the RSEM files.
                                      Rownames must be the samples.Column names must be metadata about the samples (e.g. 'sample_id','label', etc). 
                                      The 'sample_id' and 'label' columns MUST be present in the file and the values of the column 'sample_id' 
                                      must include the samples that correspond to the provided RSEM files.

          --rsem_file_suffix=value  - A string in single quotes, one of the following:
                                      'isoform.results' or 'genes.results'.
                                      
          --output=path             - A filename to save the hbadeals::hbadeals() output.
          
         --help                     - you are reading it
         
      Optionnal arguments:
          --n_cores=value           - Number of cores for hbadeals::hbadeals(), default: 2
          --isoform_level=boolean   - TRUE or FALSE. Check ?hbadeals::hbadeals, default: TRUE
          --mcmc_warmup=value       - Number of warmup iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 20
          --mcmc_iter=value         - Number of iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 100
          
     Usage:
      
          The typical command for running the script is as follows:
    
          ./rsem2hbadeals.R --rsem_folder='.' --metadata='./metadata.csv' --rsem_file_suffix='.isoforms.results' --output='hbadeals_results.csv'
      
      
      WARNING : here put all the things the user has to know
      \n")
  
  
  q(save="no")
}

## Parse arguments (we expect the form --arg=value)
parseArgs    <- function(x) strsplit(sub("^--", "", x), "=")

argsL        <- as.list(as.character(as.data.frame(do.call("rbind", parseArgs(args)))$V2))
names(argsL) <- as.data.frame(do.call("rbind", parseArgs(args)))$V1
args         <- argsL
rm(argsL)

## Give some value to optional arguments if not provided
if(is.null(args$n_cores))       {args$n_cores       = 2}    else {args$n_cores=as.numeric(args$n_cores)}
if(is.null(args$isoform_level)) {args$isoform_level = TRUE} else {args$isoform_level=as.logical(args$isoform_level)}
if(is.null(args$mcmc_warmup))   {args$mcmc_warmup   = 20}   else {args$mcmc_warmup=as.numeric(args$mcmc_warmup)}
if(is.null(args$mcmc_iter))     {args$mcmc_iter     = 100}  else {args$mcmc_iter=as.numeric(args$mcmc_iter)}

cat("\n")
cat("ARGUMENTS SUMMARY")
cat("\n")
cat("rsem_folder      : ", args$rsem_folder,    "\n",sep="")
cat("metadata         : ", args$metadata,        "\n",sep="")
cat("rsem_file_suffix : ", args$rsem_file_suffix,"\n",sep="")
cat("output           : ", args$output          ,"\n",sep="")
cat("n_cores          : ", args$n_cores,         "\n",sep="")
cat("isoform_level    : ", args$isoform_level,   "\n",sep="")
cat("mcmc_warmup      : ", args$mcmc_warmup,     "\n",sep="")
cat("mcmc_iter        : ", args$mcmc_iter,       "\n",sep="")


# Facilitates testing
rsem_folder        <- args$rsem_folder 
metadata           <- args$metadata
rsem_file_suffix   <- args$rsem_file_suffix
output             <- args$output          
n_cores            <- args$n_cores    
isoform_level      <- args$isoform_level
mcmc_warmup        <- args$mcmc_warmup    
mcmc_iter          <- args$mcmc_iter  

# # Facilitates testing
# rsem_folder        <- "."
# metadata           <- "metadata.csv"
# rsem_file_suffix   <- ".isoforms.results"
# output             <- "lalal.csv"      
# n_cores            <- 2
# isoform_level      <- TRUE
# mcmc_warmup        <- 2
# mcmc_iter          <- 4

############################## LIBRARIES SECTION #############################

suppressWarnings(suppressMessages(library(limma)))
suppressWarnings(suppressMessages(library(coda)))
suppressWarnings(suppressMessages(library(rstan)))
suppressWarnings(suppressMessages(library(codetools)))
suppressWarnings(suppressMessages(library(hbadeals)))

# ############################### SCRIPT SECTION ###############################


# Read metadata
metaData  <- read.table(metadata, header = TRUE, sep = ",")

# Get paths of RSEM files
isoform_individual_files <- list.files(rsem_folder,
                                       full.names = TRUE,
                                       pattern    = rsem_file_suffix)


# Keep 'gene_id' and 'transcript_id' columns only (scaffold to build collective countsData csv from all samples)
countsData  <- read.table(isoform_individual_files[1], header=TRUE)
countsData  <- countsData[, c("transcript_id", "gene_id")]

# Iteratively append each individual sample RSEM 'expected_count' column to create cohort table
for (input.file in isoform_individual_files  ) {
	next.file  <- read.table(input.file, header=TRUE)
	sample_id  <- gsub(rsem_file_suffix, "", basename(input.file))
	
	# Heuristic to name a column after the sample id
	next.file[[sample_id ]] <- next.file$expected_count
	toKeep     <- c("transcript_id", sample_id)
	countsData <- merge(countsData, next.file[, toKeep], by = "transcript_id")
}

# Calculating sample size
sample_size <-  dim(countsData)[2] - 2
message(paste0('Sample size = ', sample_size))

# Temporarily dettach metadata columns "transcript_id","gene_id" from numerical values

all      <- colnames(countsData)
toRemove <- c("transcript_id","gene_id")
toKeep   <- all [!all  %in% toRemove]

# Apply exclusion criterion 1: Filter out transcripts with cumsum across samples lower than the N, cohort size
message("\nApplying exclusion criterion 1:  ( discard transcripts with very low count across samples) ..")
countsData <- countsData[rowSums(countsData[,toKeep] > 0 ) >= sample_size, ]
message("Done!")

# Apply exclusion criterion 2: Filter out transcripts with only one isoform
message("\nApplying exclusion criterion 2: ( keep > 1 isoform transcripts) ..")
num.iso=unlist(lapply(countsData$gene_id, function(x){sum(countsData$gene_id %in% x)}))
countsData <- countsData[num.iso > 1, ]
message("Done!")

# Reorder countsData columns by metaData order !DANGEROUS TO RELY ON INDEXES!
message("\nOrdering 'countsData' columns according to 'metaData$sample_id' order ..")
toKeepInOrder <- c("transcript_id", "gene_id", as.vector(metaData$sample_id) )
countsData    <- countsData[, toKeepInOrder]
message("Done!")


# Run HBA-DEALS
message('\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')
# res <-hbadeals::hbadeals(countsData     = countsData,
#                           labels        = metaData$label,
#                           n.cores       = n_cores,
#                           isoform.level = isoform_level,
#                           mcmc.iter     = mcmc_iter,
#                           mcmc.warmup   = mcmc_warmup)
message('Successfully completed hbadeals::hbadeals() task!')

# Write results in file
message('\nWriting  hbadeals::hbadeals() results to a file..')
# write.table(res,
#             file = output,
#             sep       =',',
#             quote     = F,
#             col.names = T,
#             row.names = F)
message('Done!')
message('\nGoodbye!')


# TODO: Remove after Nextflow testing
system(paste0("touch ", output))


